node/src/ledger/ledger_manager.rs

```diff
@@ -0,0 +1,627 @@
+                Ok(LedgerResponse::LedgerAuxAndCoinbaseResult(res))
+            }
+            LedgerRequest::StagedLedgerDiffCreate {
+                pred_block,
+                won_slot,
+                coinbase_receiver,
+                completed_snarks,
+                supercharge_coinbase,
+            } => ledger_ctx
+                .staged_ledger_diff_create(
+                    &pred_block,
+                    &won_slot,
+                    &coinbase_receiver,
+                    completed_snarks,
+                    supercharge_coinbase,
+                )
+                .map(LedgerResponse::StagedLedgerDiff),
+            LedgerRequest::StagedLedgerReconstruct {
+                snarked_ledger_hash,
+                parts,
+            } => {
+                ledger_ctx.staged_ledger_reconstruct(snarked_ledger_hash, parts);
+                Ok(LedgerResponse::Success)
+            }
+            LedgerRequest::StagedLedgerReconstructResultStore { staged_ledger_hash } => {
+                ledger_ctx.staged_ledger_reconstruct_result_store(staged_ledger_hash);
+                Ok(LedgerResponse::Success)
+            }
+            LedgerRequest::StakeProofSparseLedger {
+                staking_ledger,
+                producer,
+                delegator,
+            } => {
+                let res = ledger_ctx.stake_proof_sparse_ledger(staking_ledger, producer, delegator);
+                Ok(LedgerResponse::SparseLedgerBase(res))
+            }
+        }
+    }
+}
+
+struct LedgerRequestWithChan {
+    request: LedgerRequest,
+    responder: Sender<Result<LedgerResponse, String>>,
+}
+
+pub struct LedgerManager {
+    sender: mpsc::UnboundedSender<LedgerRequestWithChan>,
+    runtime: thread::JoinHandle<LedgerCtx>,
+}
+
+impl LedgerManager {
+    pub fn new(ledger_ctx: LedgerCtx) -> LedgerManager {
+        let (sender, receiver) = mpsc::unbounded_channel();
+        let runtime = thread::spawn(move || {
+            let mut ledger_ctx = ledger_ctx;
+            let mut receiver = receiver;
+            loop {
+                match receiver.blocking_recv() {
+                    Some(LedgerRequestWithChan { request, responder }) => responder
+                        .send(request.handle(&mut ledger_ctx))
+                        .unwrap_or(()),
+                    None => {
+                        break;
+                    }
+                }
+            }
+            ledger_ctx
+        });
+        LedgerManager { sender, runtime }
+    }
+
+    fn call(&self, request: LedgerRequest) -> Result<LedgerResponse, String> {
+        let (responder, receiver) = channel();
+        self.sender
+            .send(LedgerRequestWithChan { request, responder })
+            .unwrap();
+        result_join(receiver.recv())
+    }
+
+    pub async fn wait_for_stop(
+        self,
+    ) -> Result<LedgerCtx, Box<dyn std::any::Any + std::marker::Send + 'static>> {
```

Comment:
Yes this function is indeed unused at the moment. And yes, the thread ends in case the handle gets dropped, which at the moment is when the whole `NodeService` is dropped. I added the function for completeness of the API and also in case someone needed to stop the manager explicitly at some point and retrieve the LedgerCtx. Admittedly, I cannot imagine a scenario in which that would be better than addding the necessary functionality to the manager, but then again I don't know the system that well yet. In any case, I can remove it if you prefer.

Reply:
@Sventimir leave it there for now. At some point we will have to join that thread because it may be in the middle of some I/O operations when the node gets shut down that we may want to wait for (specially if we implement a persisted version of the ledger, scan state, etc to lower memory usage)

Reply:
Okay, leaving it there then.